pkg/ingress/controller/endpointslice.go

```diff
@@ -0,0 +1,251 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/apache/apisix-ingress-controller/pkg/apisix"
+	apisixcache "github.com/apache/apisix-ingress-controller/pkg/apisix/cache"
+	"github.com/apache/apisix-ingress-controller/pkg/kube"
+	"github.com/apache/apisix-ingress-controller/pkg/log"
+	"github.com/apache/apisix-ingress-controller/pkg/seven/state"
+	"github.com/apache/apisix-ingress-controller/pkg/types"
+	apisixv1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
+)
+
+type endpointSliceController struct {
+	controller *Controller
+	workqueue  workqueue.RateLimitingInterface
+	workers    int
+	epInformer cache.SharedIndexInformer
+}
+
+func (c *Controller) newEndpointSliceController() *endpointSliceController {
+	ctl := &endpointSliceController{
+		controller: c,
+		workqueue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "endpointslice"),
+		workers:    1,
+		epInformer: kube.CoreSharedInformerFactory.Discovery().V1beta1().EndpointSlices().Informer(),
+	}
+
+	ctl.epInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctl.onAdd,
+			UpdateFunc: ctl.onUpdate,
+			DeleteFunc: ctl.onDelete,
+		},
+	)
+
+	return ctl
+}
+
+func (c *endpointSliceController) run(ctx context.Context) {
+	log.Info("endpointslice controller started")
+	defer log.Info("endpointslice controller exited")
+
+	if ok := cache.WaitForCacheSync(ctx.Done(), c.epInformer.HasSynced); !ok {
+		log.Error("informers sync failed")
+		return
+	}
+
+	handler := func() {
+		for {
+			obj, shutdown := c.workqueue.Get()
+			if shutdown {
+				return
+			}
+
+			err := c.sync(ctx, obj.(*types.Event))
+			c.workqueue.Done(obj)
+			c.handleSyncErr(obj, err)
+		}
+	}
+
+	for i := 0; i < c.workers; i++ {
+		go handler()
+	}
+
+	<-ctx.Done()
+	c.workqueue.ShutDown()
+}
+
+func (c *endpointSliceController) sync(ctx context.Context, ev *types.Event) error {
+	ep := ev.Object.(*discoveryv1beta1.EndpointSlice)
+	svc, err := c.controller.svcLister.Services(ep.Namespace).Get(ep.Name)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Warnf("service %s/%s was deleted", ep.Namespace, ep.Name)
+			return nil
+		}
+		log.Errorf("failed to get service %s/%s: %s", ep.Namespace, ep.Name, err)
+		return err
+	}
+	portMap := make(map[string]int32)
+	for _, port := range svc.Spec.Ports {
+		portMap[port.Name] = port.Port
+	}
+	clusters := c.controller.apisix.ListClusters()
+	for _, port := range ep.Ports {
+		svcPort, ok := portMap[*port.Name]
+		if !ok {
+			// This shouldn't happen.
+			log.Errorf("port %s in endpointslice %s/%s but not in service", *port.Name, ep.Namespace, ep.Name)
+			continue
+		}
+		nodes, err := c.controller.translator.TranslateUpstreamNodesFromEndpointSlice(ep, svcPort)
+		if err != nil {
+			log.Errorw("failed to translate upstream nodes",
+				zap.Error(err),
+				zap.Any("endpointslice", ep),
+				zap.Int32("port", svcPort),
+			)
```

Comment:
It seems we should return `err` here since error occurred. The same logic in endpoints controller should also be changed.